src/c_api/c_api_ndarray.cc

```diff
@@ -798,32 +341,58 @@ int MXAutogradBackward(mx_uint num_output,
                        NDArrayHandle *output_handles,
                        NDArrayHandle *ograd_handles,
                        int retain_graph) {
-  return MXAutogradBackwardEx(num_output, output_handles, ograd_handles, retain_graph, true);
+  return MXAutogradBackwardEx(num_output, output_handles, ograd_handles,
+                              0, nullptr, retain_graph, false, true,
+                              nullptr, nullptr);
 }
 
 int MXAutogradBackwardEx(mx_uint num_output,
                          NDArrayHandle *output_handles,
                          NDArrayHandle *ograd_handles,
+                         mx_uint num_variables,
+                         NDArrayHandle *var_handles,
                          int retain_graph,
-                         int is_train) {
+                         int create_graph,
+                         int is_train,
+                         NDArrayHandle **grad_handles,
+                         int **grad_stypes) {
+  MXAPIThreadLocalEntry *ret = MXAPIThreadLocalStore::Get();
   API_BEGIN();
 
-  std::vector<NDArray> outputs, ograds;
+  std::vector<NDArray*> outputs, ograds, variables;
   outputs.reserve(num_output);
   for (mx_uint i = 0; i < num_output; ++i) {
-    outputs.emplace_back(*static_cast<NDArray*>(output_handles[i]));
+    outputs.emplace_back(reinterpret_cast<NDArray*>(output_handles[i]));
   }
 
   ograds.reserve(num_output);
   for (mx_uint i = 0; i < num_output; ++i) {
-    if (ograd_handles != nullptr && ograd_handles[i] != nullptr) {
-      ograds.emplace_back(*static_cast<NDArray*>(ograd_handles[i]));
+    if (ograd_handles != nullptr) {
+      ograds.emplace_back(reinterpret_cast<NDArray*>(ograd_handles[i]));
     } else {
-      ograds.emplace_back();
+      ograds.emplace_back(nullptr);
```

Comment:
`ograds.emplace_back();` always generates dense ograd. Does `nullptr` respect the storage type inferred from the graph?